simplify run(), deprecate populate_multiple_folders()

iantaylor-NOAA opened the pull request:
This issue resolves #733 by removing the functionality of having `run()` work on a vector of directories and deprecates the function `populate_multiple_folders()`. I think it's better practice in general to keep these functions smaller and rely on packages like {purrr} to apply them iteratively. We may also want to explore {furrr} for parallel processing with `run()`: https://furrr.futureverse.org/.

kellijohnson-NOAA commented on R/run.R:
```diff
@@ -74,82 +78,75 @@ run <- function(dir = getwd(),
-  # return table of results
-  return(data.frame(dir = dir, results = results))
+      }
+      if (isTRUE(console_output > 0)) {
+        results <- "model run failed"
+      } else if (isTRUE(console_output == 0)) {
+        results <- "ran model"
+      } else {
+        results <- "unknown run status"
+      }
```
I am not positive but I think you could change this complicated chain of if else statements to the following
```
results <- dplyr::case_when(
  console_ouput == 0 ~ "ran model",
  console_output > 0 ~ "model run failed",
  "unknown run status"
)
```

iantaylor-NOAA replied:
Good idea. Looks like final line needs to be `TRUE ~ "unknown run status"`

kellijohnson-NOAA reviewed (changes requested):
Most comments are related to documentation and condensing the code. The former is more important than the latter imo.

iantaylor-NOAA commented on R/populate_multiple_folders.R:
```diff
@@ -39,6 +44,15 @@ populate_multiple_folders <- function(outerdir.old = NULL,
+    what = "populate_multiple_folders()",
+    details = paste(
+      "This function has been deprecated.",
+      "Please use r4ss::copy_SS_inputs() along with {purrr} or other tools."
```
Thanks for the purrr and lapply examples. I suspect that nobody is using this function and it will be deleted before long, so I'm going to focus the example-writing on the documentation of copy_SS_inputs() where it will have a more lasting impact.

kellijohnson-NOAA commented:
One more thought I had was should we get rid of the test for the function if the function is still available to users?

iantaylor-NOAA commented:
@kellijohnson-NOAA, thanks for the really helpful review. Sorry I left so many legacy elements in `run()` that should have been gone before this PR was created.
This morning I spent too long trying to create a good example of how to replace `populate_multiple_folders()` that was equally useful and easy and finally gave up and reverted the commit to deprecate that function. Instead I've cleaned up the function to work better with `copy_SS_inputs()` which also needed some cleanup.

If you have time for another look at the changes, your input is appreciated.

kellijohnson-NOAA commented on R/run.R:
```diff
@@ -34,122 +30,120 @@
+    warning("not a directory:", dir)
+    results <- "not a directory"
+  } else {
+    if (skipfinished & "Report.sso" %in% dir(dir)) {
```
I think I might have mentioned this before but I believe this should be
```
if (file.exists(file.path(dir,"Report.sso")) && skipfinished) {
```
because you only care about skipfinished if the file actually exists. Furthermore, I think that using file.exists rather than %in% is more robust.

kellijohnson-NOAA commented on R/run.R:
```diff
@@ -34,122 +30,120 @@
-  } # end loop over directories
-  # return table of results
-  return(data.frame(dir = dir, results = results))
+      setwd(wd_orig) # needed when using relative paths
```
This seems redundant with line 88 and line 88 is more robust.

kellijohnson-NOAA commented on man-roxygen/show_in_console.R:
```diff
@@ -1,2 +1,2 @@
 #' @param show_in_console Show output in the R console? If FALSE,
-#' then it is saved to a file.
+#' then it is saved to a file at the end of the model run.
```
Do you  need to have this in a template given that it is only used in a single function, i.e., `run()`. Also, there is no default argument called out in the explanation. Line two should be indented by two spaces. FALSE should be wrapped in back ticks. It is not 100% clear what "it" on line two is.

kellijohnson-NOAA commented on R/populate_multiple_folders.R:
```diff
@@ -11,10 +11,12 @@
-#' @param exe.only Only copy exe files from exe.dir, don't copy input files
 #' @template verbose
 #' @return Returns table of results indicating which directories were
 #' successfully populated with the model input files and/or executables
```
Full stop needed at end of line 22.

kellijohnson-NOAA reviewed (changes requested):
😆 I did the exact same thing in {ss3sim} where I thought I was going to deprecate something and then it ended up being more work than it was worth. I reviewed the new changes and most of them look good to me except the new argument in `population_mutliple_folders()`. I am not sure that it is needed and if you do decide to keep it, I think it needs to be documented a little bit more thoroughly so users know how it interacts with `exe.dir`. Personally, I think that it can be removed.
I am requesting changes largely because I would like to see the examples actually work on people's machines. As is the example for `populate_multiple_folders()` does not work. I provided code that might work but I am unsure if the `outdir.new` does not already exist, if it will work? Perhaps this leads to an additional comment that `outerdir.new` should be documented more. Users should know if this location needs to already exist.

iantaylor-NOAA commented on R/populate_multiple_folders.R:
```diff
@@ -11,10 +11,12 @@
 #' @template overwrite
 #' @param use_ss_new Use .ss_new files instead of original inputs?
-#' @param exe.dir Path to executable to copy into all the subfolders.
+#' @param copy_exe Copy any executables found in the individual
+#' directories or in exe.dir (if provided)? To copy a new executable
+#' from a central location, use `exe.dir` instead.
```
This is a great solution. I was not happy with the two arguments but couldn't think of a better solution like the one you propose.

iantaylor-NOAA commented on R/populate_multiple_folders.R:
```diff
@@ -26,28 +28,39 @@
 #' populate_multiple_folders(
 #'   outerdir.old = "c:/SS/old_models",
 #'   outerdir.new = "c:/SS/new_models",
```
Good idea. The `creatdir = TRUE` argument allows it to create the new directory created by tempdir()

iantaylor-NOAA commented:
Thanks for the additional input. These functions are now simpler, better documented, and with better examples.

iantaylor-NOAA commented:
@kellijohnson-NOAA, I'm going to go ahead and merge this in spite of being on leave for the next week. I think that the changes fix more things than they are likely to break if I made some additional error and would like to get the improved documentation available to people starting to use the revised functions that changed in #723.

In spite of all the work that this took, I think there will be very little impact on users because `populate_multiple_folders()` is not really used by anyone yet, nor is the option to pass a vector of directories to `run()`.

kellijohnson-NOAA commented:
thanks @iantaylor-NOAA for all of your work on this. If it makes you feel any better, {ss3sim} now uses `r4ss::run()` 😉 as a result of it.
